### PR TITLE
FIX: force a read operation for peer instead of self

### PR DESCRIPTION
**Pull request by louxiu**
Motivation:
When A is in `writeInProgress` and call self close, A should
`finishPeerRead` for B(A' peer).

Modifications:
Call `finishPeerRead` with peer in `LocalChannel#doClose`

Result:
Clear confuse of code logic

### CODE REVIEW

**Comment by louxiu** [comment]
This commit doesn't fix any bug for LocalChannel. IMHO, the original code is just incorrect. 
@normanmaurer @Scottmitch  PTAL

**Comment by louxiu** [comment]
Add another [commit](https://github.com/netty/netty/pull/7454/commits/8849a3ae8a2e2808742aa672ed9500e2ab332002) to solve an order problem in same event group.

**Comment by normanmaurer** [comment]
@louxiu please check the test failure

**Comment by louxiu** [comment]
@normanmaurer done.  This commit improves mismatch test in `SniClientTest `.

File: transport/src/test/java/io/netty/channel/local/LocalChannelTest.java [hunk 1]
```diff
@@ -390,6 +391,67 @@ public void operationComplete(ChannelFuture future) throws Exception {
...
+                    .channel(LocalChannel.class)
+                    .handler(new ChannelInboundHandlerAdapter() {
+                        @Override
+                        public void channelActive(ChannelHandlerContext ctx) throws Exception {
+                            ctx.writeAndFlush(data.retainedDuplicate());
+                        }
+
+                        @Override
+                        public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+                            assertEquals(data, msg);
```
Comment (Scottmitch) [inline transport/src/test/java/io/netty/channel/local/LocalChannelTest.java]: consider surfacing this exception in the JUnit thread. One option is to set an AtomicReference<Throwable> and check it after the latch await returns.

**Comment by louxiu** [comment]
@Scottmitch Done it another way. PTAL

**Pull request merged**